### PR TITLE
hal/ia32: Add basic ACPI support

### PR DESCRIPTION
**Pull request by astalke**
## Motivation and Context

Kernel needs ACPI for more advanced features. This PR gives the kernel access to MADT and FADT tables.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: ia32-generic-qemu

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [x] This PR needs additional PRs to work: https://github.com/phoenix-rtos/phoenix-rtos-kernel/pull/464
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: hal/ia32/acpi.c [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+{
+	static const char RSDP_MAGIC[] = "RSD PTR ";
+	void *ptr;
+	rsdp_t *entry;
+	for (ptr = start; ptr < end; ptr += 16) {
+		if (hal_strncmp(RSDP_MAGIC, ptr, 8) == 0) {
+			if (_hal_checksumVerify(ptr, sizeof(rsdp_t)) != 0) {
+				entry = ptr;
+				if ((entry->revision == 2) && (_hal_checksumVerify(entry, sizeof(xsdp_t)) != 0)) {
+					hs->acpi_version = ACPI_XSDP;
```
Comment (agkaminski) [inline hal/ia32/acpi.c:87]: Where does the `ACPI_XSDP` come from? (looks like something hasn't been staged to commit)

File: hal/ia32/acpi.c [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+				}
+				else {
+					hs->acpi_version = ACPI_RSDP;
```
Comment (agkaminski) [inline hal/ia32/acpi.c:90]: Same, `ACPI_RSDP` isn't defined

File: hal/ia32/memory.c [hunk 3]
```diff
@@ -29,6 +29,8 @@ extern char __bss_start[], __bss_end[];
 extern char __heap_base[], __heap_limit[];
 extern char __stack_top[], __stack_limit[];
 
+extern void hal_acpiInit(hal_syspage_t *);
```
Comment (agkaminski) [inline hal/ia32/memory.c:32]: I know it's just one function prototype, but it would be neater to provide a header file for ACPI

File: syspage.c [hunk 4]
```diff
@@ -539,24 +579,29 @@ void syspage_mapShow(void)
...
 
 static void syspage_progMapsShow(const syspage_prog_t *prog)
 {
-	unsigned int i;
+	size_t i;
 
-	for (i = 0; i < prog->imapSz || i < prog->dmapSz; ++i) {
-		if (i)
+	for (i = 0; (i < prog->imapSz) || (i < prog->dmapSz); ++i) {
+		if (i) {
```
Comment (agkaminski) [inline syspage.c:592]:
```suggestion
		if (i != 0) {
```

File: hal/ia32/acpi.c (hunk 1 above)
Comment (astalke) [reply hal/ia32/acpi.c:87]: They're included outside of plo in: `phoenix-rtos-kernel/include/arch/syspage-ia32.h`

File: hal/ia32/acpi.c [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+	hs->fadt = 0;
+	hs->fadtLength = 0;
+
+	rsdp_t *rsdp = _hal_acpiFindRsdp(hs);
+	if (rsdp != NULL) {
+		switch (hs->acpi_version) {
+			case ACPI_XSDP:
+				/* TODO: Don't fallback to RSDP */
+				hs->acpi_version = ACPI_RSDP;
+			case ACPI_RSDP:
```
Comment (agkaminski) [inline hal/ia32/acpi.c:168]: Missing break or intentional? Add `fall through` comment if intentional

File: hal/ia32/acpi.h [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+ * Copyright 2023 Phoenix Systems
+ * Author: Andrzej Stalke
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+#include <hal/hal.h>
+
+void hal_acpiInit(hal_syspage_t *hs);
```
Comment (agkaminski) [inline hal/ia32/acpi.h:17]: Missing header guard

File: syspage.c [hunk 7]
```diff
@@ -140,30 +144,30 @@ static int syspage_strAttr2ui(const char *str, unsigned int *attr)
...
 				break;
 			case mAttrBufferable:
-				str[pos++] = 'b';
+				str[pos] = 'b';
 				break;
 		}
+		++pos;
 	}
 
 	str[pos] = '\0';
```
Comment (kemonats) [inline syspage.c:173]: Your implementation is not good.
Simple example:
`attr == 1; // mAttrRead`
`str == "uruuuuuuuuuuuuuuuuuuuuuuuuuuuuuu\0"`
where `u` means undefined
Leave it as it is now or change it to : 
```c
{
	char *p = str;

	if ((attr & mAttrRead) != 0) {
		*p++ = 'r';
	}

	if ((attr & mAttrWrite) != 0) {
		*p++ = 'w';
	}

	if ((attr & mAttrExec) != 0) {
		*p++ = 'x';
	}

	if ((attr & mAttrShareable) != 0) {
		*p++ = 's';
	}

	if ((attr & mAttrCacheable) != 0) {
		*p++ = 'c';
	}

	if ((attr & mAttrBufferable) != 0) {
		*p++ = 'b';
	}

	*p = '\0';
}
```

Comment (astalke) [reply syspage.c]: I'll revert it back.

**Review by agkaminski — APPROVED** [review]
LGTM

**Pull request merged**